File: python_programs_and_containers/basic_container_lite/libraries/file_loader/test_files/kb_go/postgres/data_structures/data_structures_module/KBSearch.go
package data_structures_module

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"

	//"github.com/lib/pq"
	//_ "github.com/lib/pq"
)

// Filter represents a query filter with condition and parameters
type Filter struct {
	Condition string
	Params    map[string]interface{}
}

// KBSearch handles SQL filtering for the knowledge_base table
type KBSearch struct {
	Path           []string
	Host           string
	Port           string
	DBName         string
	User           string
	Password       string
	BaseTable      string
	LinkTable      string
	LinkMountTable string
	Filters        []Filter
	Results        []map[string]interface{}
	PathValues     map[string]interface{}
	conn           *sql.DB
}

// NewKBSearch creates a new KBSearch instance and connects to the database
func NewKBSearch(host, port, dbname, user, password, database string) (*KBSearch, error) {
	kb := &KBSearch{
		Path:           []string{},
		Host:           host,
		Port:           port,
		DBName:         dbname,
		User:           user,
		Password:       password,
		BaseTable:      database,
		LinkTable:      database + "_link",
		LinkMountTable: database + "_link_mount",
		Filters:        []Filter{},
		PathValues:     make(map[string]interface{}),
	}

	if err := kb.connect(); err != nil {
		return nil, err
	}

	return kb, nil
}

// connect establishes a connection to the PostgreSQL database
func (kb *KBSearch) connect() error {
	connStr := fmt.Sprintf("host=%s port=%s dbname=%s user=%s password=%s sslmode=disable",
		kb.Host, kb.Port, kb.DBName, kb.User, kb.Password)

	conn, err := sql.Open("postgres", connStr)
	if err != nil {
		return fmt.Errorf("error connecting to database: %v", err)
	}

	// Test the connection
	if err := conn.Ping(); err != nil {
		return fmt.Errorf("error pinging database: %v", err)
	}

	kb.conn = conn
	return nil
}

// Disconnect closes the database connection
func (kb *KBSearch) Disconnect() error {
	if kb.conn != nil {
		return kb.conn.Close()
	}
	return nil
}

// GetConnAndCursor returns the database connection
func (kb *KBSearch) GetConnAndCursor() (*sql.DB, error) {
	if kb.conn == nil {
		return nil, fmt.Errorf("not connected to database. Call connect() first")
	}
	return kb.conn, nil
}

// ClearFilters clears all filters and resets the query state
func (kb *KBSearch) ClearFilters() {
	kb.Filters = []Filter{}
	kb.Results = nil
}

// SearchKB adds a filter to search for rows matching the specified knowledge_base
func (kb *KBSearch) SearchKB(knowledgeBase string) {
	kb.Filters = append(kb.Filters, Filter{
		Condition: "knowledge_base = $knowledge_base",
		Params:    map[string]interface{}{"knowledge_base": knowledgeBase},
	})
}

// SearchLabel adds a filter to search for rows matching the specified label
func (kb *KBSearch) SearchLabel(label string) {
	kb.Filters = append(kb.Filters, Filter{
		Condition: "label = $label",
		Params:    map[string]interface{}{"label": label},
	})
}

// SearchName adds a filter to search for rows matching the specified name
func (kb *KBSearch) SearchName(name string) {
	kb.Filters = append(kb.Filters, Filter{
		Condition: "name = $name",
		Params:    map[string]interface{}{"name": name},
	})
}

// SearchPropertyKey adds a filter to search for rows where properties contains the key
func (kb *KBSearch) SearchPropertyKey(key string) {
	kb.Filters = append(kb.Filters, Filter{
		Condition: "properties::jsonb ? $property_key",
		Params:    map[string]interface{}{"property_key": key},
	})
}

// SearchPropertyValue adds a filter to search for rows where properties contains the key-value pair
func (kb *KBSearch) SearchPropertyValue(key string, value interface{}) {
	jsonObject := map[string]interface{}{key: value}
	jsonBytes, _ := json.Marshal(jsonObject)

	kb.Filters = append(kb.Filters, Filter{
		Condition: "properties::jsonb @> $json_object::jsonb",
		Params:    map[string]interface{}{"json_object": string(jsonBytes)},
	})
}

// SearchStartingPath adds a filter to search for descendants of the specified path
func (kb *KBSearch) SearchStartingPath(startingPath string) {
	kb.Filters = append(kb.Filters, Filter{
		Condition: "path <@ $starting_path",
		Params:    map[string]interface{}{"starting_path": startingPath},
	})
}

// SearchPath adds a filter to search for rows matching the LTREE path expression
func (kb *KBSearch) SearchPath(pathExpression string) {
	kb.Filters = append(kb.Filters, Filter{
		Condition: "path ~ $path_expr",
		Params:    map[string]interface{}{"path_expr": pathExpression},
	})
}

// SearchHasLink adds a filter to search for rows where has_link is TRUE
func (kb *KBSearch) SearchHasLink() {
	kb.Filters = append(kb.Filters, Filter{
		Condition: "has_link = TRUE",
		Params:    map[string]interface{}{},
	})
}

// SearchHasLinkMount adds a filter to search for rows where has_link_mount is TRUE
func (kb *KBSearch) SearchHasLinkMount() {
	kb.Filters = append(kb.Filters, Filter{
		Condition: "has_link_mount = TRUE",
		Params:    map[string]interface{}{},
	})
}

// ExecuteQuery executes the progressive query with all added filters using CTEs
func (kb *KBSearch) ExecuteQuery() ([]map[string]interface{}, error) {
	if kb.conn == nil {
		return nil, fmt.Errorf("not connected to database")
	}

	columnStr := "*"

	// If no filters, execute simple query
	if len(kb.Filters) == 0 {
		query := fmt.Sprintf("SELECT %s FROM %s", columnStr, kb.BaseTable)
		rows, err := kb.conn.Query(query)
		if err != nil {
			return nil, err
		}
		defer rows.Close()

		results, err := kb.rowsToMaps(rows)
		if err != nil {
			return nil, err
		}
		kb.Results = results
		return results, nil
	}

	// Build CTE query
	cteParts := []string{}
	paramSlice := []interface{}{}
	paramCounter := 1

	// Initial CTE
	cteParts = append(cteParts, fmt.Sprintf("base_data AS (SELECT %s FROM %s)", columnStr, kb.BaseTable))

	// Process each filter
	for i, filter := range kb.Filters {
		condition := filter.Condition
		params := filter.Params

		// Replace parameter placeholders with positional parameters
		for paramName, paramValue := range params {
			placeholder := "$" + paramName
			newPlaceholder := fmt.Sprintf("$%d", paramCounter)
			condition = strings.Replace(condition, placeholder, newPlaceholder, -1)
			paramSlice = append(paramSlice, paramValue)
			paramCounter++
		}

		cteName := fmt.Sprintf("filter_%d", i)
		prevCTE := "base_data"
		if i > 0 {
			prevCTE = fmt.Sprintf("filter_%d", i-1)
		}

		var cteQuery string
		if condition != "" && len(params) > 0 {
			cteQuery = fmt.Sprintf("%s AS (SELECT %s FROM %s WHERE %s)", cteName, columnStr, prevCTE, condition)
		} else if condition != "" {
			cteQuery = fmt.Sprintf("%s AS (SELECT %s FROM %s WHERE %s)", cteName, columnStr, prevCTE, condition)
		} else {
			cteQuery = fmt.Sprintf("%s AS (SELECT %s FROM %s)", cteName, columnStr, prevCTE)
		}

		cteParts = append(cteParts, cteQuery)
	}

	// Build final query
	withClause := "WITH " + strings.Join(cteParts, ",\n")
	finalSelect := fmt.Sprintf("SELECT %s FROM filter_%d", columnStr, len(kb.Filters)-1)
	finalQuery := fmt.Sprintf("%s\n%s", withClause, finalSelect)

	// Execute query
	rows, err := kb.conn.Query(finalQuery, paramSlice...)
	if err != nil {
		return nil, fmt.Errorf("error executing query: %v\nQuery: %s\nParams: %v", err, finalQuery, paramSlice)
	}
	defer rows.Close()

	results, err := kb.rowsToMaps(rows)
	if err != nil {
		return nil, err
	}

	kb.Results = results
	return results, nil
}

// rowsToMaps converts SQL rows to slice of maps
func (kb *KBSearch) rowsToMaps(rows *sql.Rows) ([]map[string]interface{}, error) {
	columns, err := rows.Columns()
	if err != nil {
		return nil, err
	}

	results := []map[string]interface{}{}

	for rows.Next() {
		// Create a slice of interface{} to hold the values
		values := make([]interface{}, len(columns))
		valuePointers := make([]interface{}, len(columns))

		for i := range columns {
			valuePointers[i] = &values[i]
		}

		if err := rows.Scan(valuePointers...); err != nil {
			return nil, err
		}

		// Create a map for this row
		entry := make(map[string]interface{})
		for i, col := range columns {
			var v interface{}
			val := values[i]
			b, ok := val.([]byte)
			if ok {
				v = string(b)
			} else {
				v = val
			}
			entry[col] = v
		}

		results = append(results, entry)
	}

	return results, nil
}

// FindPathValues extracts path values from query results
func (kb *KBSearch) FindPathValues(keyData []map[string]interface{}) []string {
	if len(keyData) == 0 {
		return []string{}
	}

	returnValues := []string{}
	for _, row := range keyData {
		if path, ok := row["path"].(string); ok {
			returnValues = append(returnValues, path)
		}
	}

	return returnValues
}

// GetResults returns the results of the last executed query
func (kb *KBSearch) GetResults() []map[string]interface{} {
	if kb.Results == nil {
		return []map[string]interface{}{}
	}
	return kb.Results
}




// FindDescription extracts description from properties field of a single query result
func (kb *KBSearch) FindDescription(row map[string]interface{}) map[string]string {
	description := ""
	path := ""

	// Extract path
	if p, ok := row["path"].(string); ok {
		path = p
	}

	// Extract description from properties
	if props, ok := row["properties"]; ok {
		// Handle properties as JSON
		var propsMap map[string]interface{}
		switch p := props.(type) {
		case string:
			json.Unmarshal([]byte(p), &propsMap)
		case map[string]interface{}:
			propsMap = p
		}

		if desc, ok := propsMap["description"].(string); ok {
			description = desc
		}
	}

	return map[string]string{path: description}
}

// FindDescriptions extracts descriptions from properties field of multiple query results
func (kb *KBSearch) FindDescriptions(dataSlice []map[string]interface{}) []map[string]string {
	returnValues := []map[string]string{}

	for _, row := range dataSlice {
		description := ""
		path := ""

		// Extract path
		if p, ok := row["path"].(string); ok {
			path = p
		}

		// Extract description from properties
		if props, ok := row["properties"]; ok {
			// Handle properties as JSON
			var propsMap map[string]interface{}
			switch p := props.(type) {
			case string:
				json.Unmarshal([]byte(p), &propsMap)
			case map[string]interface{}:
				propsMap = p
			}

			if desc, ok := propsMap["description"].(string); ok {
				description = desc
			}
		}

		returnValues = append(returnValues, map[string]string{path: description})
	}

	return returnValues
}

// FindDescriptionPath finds data for a single specified path in the knowledge base
func (kb *KBSearch) FindDescriptionPath(path string) (map[string]interface{}, error) {
	if path == "" {
		return map[string]interface{}{}, nil
	}

	returnValues := make(map[string]interface{})

	query := fmt.Sprintf("SELECT path, data FROM %s WHERE path = $1", kb.BaseTable)
	rows, err := kb.conn.Query(query, path)
	if err != nil {
		return nil, fmt.Errorf("error retrieving data for path: %v", err)
	}
	defer rows.Close()

	found := false
	for rows.Next() {
		var dbPath string
		var data interface{}
		if err := rows.Scan(&dbPath, &data); err != nil {
			return nil, err
		}
		returnValues[dbPath] = data
		found = true
	}

	// Add nil for path not found
	if !found {
		returnValues[path] = nil
	}

	return returnValues, nil
}

// FindDescriptionPaths finds data for multiple specified paths in the knowledge base
func (kb *KBSearch) FindDescriptionPaths(paths []string) ([]map[string]interface{}, error) {
	if len(paths) == 0 {
		return []map[string]interface{}{}, nil
	}

	returnValues := []map[string]interface{}{}

	var rows *sql.Rows
	var err error

	if len(paths) == 1 {
		// Single path optimization
		query := fmt.Sprintf("SELECT path, data FROM %s WHERE path = $1", kb.BaseTable)
		rows, err = kb.conn.Query(query, paths[0])
	} else {
		// Build query with multiple placeholders
		placeholders := make([]string, len(paths))
		args := make([]interface{}, len(paths))
		for i, path := range paths {
			placeholders[i] = fmt.Sprintf("$%d", i+1)
			args[i] = path
		}

		query := fmt.Sprintf("SELECT path, data FROM %s WHERE path IN (%s)",
			kb.BaseTable, strings.Join(placeholders, ","))
		rows, err = kb.conn.Query(query, args...)
	}

	if err != nil {
		return nil, fmt.Errorf("error retrieving data for paths: %v", err)
	}
	defer rows.Close()

	foundPaths := make(map[string]bool)

	for rows.Next() {
		var path string
		var data interface{}
		if err := rows.Scan(&path, &data); err != nil {
			return nil, err
		}
		returnValues = append(returnValues, map[string]interface{}{path: data})
		foundPaths[path] = true
	}

	// Add nil for paths not found
	for _, path := range paths {
		if !foundPaths[path] {
			returnValues = append(returnValues, map[string]interface{}{path: nil})
		}
	}

	return returnValues, nil
}

// DecodeLinkNodes decodes an ltree path into knowledge base name and node link/name pairs
func (kb *KBSearch) DecodeLinkNodes(path string) (string, [][]string, error) {
	if path == "" {
		return "", nil, fmt.Errorf("path must be a non-empty string")
	}

	pathParts := strings.Split(path, ".")

	// Need at least kb.link.name
	if len(pathParts) < 3 {
		return "", nil, fmt.Errorf("path must have at least 3 elements (kb.link.name), got %d", len(pathParts))
	}

	// After removing kb, remaining elements should be even
	remainingParts := len(pathParts) - 1
	if remainingParts%2 != 0 {
		return "", nil, fmt.Errorf("bad path format: after kb identifier, must have even number of elements (link/name pairs), got %d elements", remainingParts)
	}

	// Extract knowledge base name
	kbName := pathParts[0]

	// Process pairs of (node_link, node_name)
	result := [][]string{}
	for i := 1; i < len(pathParts); i += 2 {
		nodeLink := pathParts[i]
		nodeName := pathParts[i+1]
		result = append(result, []string{nodeLink, nodeName})
	}

	return kbName, result, nil
}

File: zig_programs_and_containers/building_blocks/knowledge_base/postgres/data_structures/luajit/kb_search.lua
--[[
  KB_Search - Knowledge Base Search with CTE-based progressive filtering
  LuaJIT port using luadbi-postgresql and dkjson.

  Usage:
    local KB_Search = require("kb_search")
    local kb = KB_Search.new({
      host     = "127.0.0.1",
      port     = 5432,
      dbname   = "mydb",
      user     = "postgres",
      password = "secret",
      database = "knowledge_base",
    })
    kb:search_kb("my_kb")
    kb:search_label("KB_STATUS_FIELD")
    local rows = kb:execute_query()
    kb:disconnect()
]]

local DBI    = require("DBI")
local dkjson = require("dkjson")

---------------------------------------------------------------------------
-- Helpers
---------------------------------------------------------------------------

--- Escape a string value for safe SQL interpolation (single-quote wrapping).
local function escape_literal(val)
  if val == nil then return "NULL" end
  local s = tostring(val)
  s = s:gsub("'", "''")
  return "'" .. s .. "'"
end

--- Fetch all rows from an executed statement as an array of dicts.
-- luadbi-postgresql: sth:fetch(true) returns a single table {col=val,...} or nil.
local function fetch_all_dicts(sth)
  local rows = {}
  while true do
    local row = sth:fetch(true)
    if not row then break end
    -- DBI fetch(true) returns the row as a table {colname=value, ...}
    -- We must copy it because DBI may reuse the table on next fetch.
    local copy = {}
    for k, v in pairs(row) do copy[k] = v end
    rows[#rows + 1] = copy
  end
  return rows
end

--- Fetch a single row as a dict, or nil.
local function fetch_one_dict(sth)
  local row = sth:fetch(true)
  if not row then return nil end
  local copy = {}
  for k, v in pairs(row) do copy[k] = v end
  return copy
end

---------------------------------------------------------------------------
-- Class
---------------------------------------------------------------------------

local KB_Search = {}
KB_Search.__index = KB_Search

function KB_Search.new(opts)
  assert(opts.host,     "host required")
  assert(opts.dbname,   "dbname required")
  assert(opts.user,     "user required")
  assert(opts.password, "password required")
  assert(opts.database, "database (table name) required")

  local self = setmetatable({}, KB_Search)
  self.host       = opts.host
  self.port       = opts.port or 5432
  self.dbname     = opts.dbname
  self.user       = opts.user
  self.password   = opts.password
  self.base_table       = opts.database
  self.link_table       = opts.database .. "_link"
  self.link_mount_table = opts.database .. "_link_mount"

  self.filters = {}
  self.results = nil
  self.path    = {}

  local dsn = string.format("dbname=%s host=%s port=%s",
    self.dbname, self.host, self.port)
  local dbh, err = DBI.Connect("PostgreSQL", dsn, self.user, self.password)
  if not dbh then
    error("Error connecting to database: " .. tostring(err))
  end
  dbh:autocommit(false)
  self.dbh = dbh

  return self
end

function KB_Search:disconnect()
  if self.dbh then
    self.dbh:close()
    self.dbh = nil
  end
end

function KB_Search:get_connection()
  if not self.dbh then
    error("Not connected to database. Call new() first.")
  end
  return self.dbh
end

function KB_Search:clear_filters()
  self.filters = {}
  self.results = nil
end

---------------------------------------------------------------------------
-- Filter builders
---------------------------------------------------------------------------

function KB_Search:search_kb(knowledge_base)
  self.filters[#self.filters + 1] = {
    condition = "knowledge_base = $knowledge_base",
    params    = { knowledge_base = knowledge_base },
  }
end

function KB_Search:search_label(label)
  self.filters[#self.filters + 1] = {
    condition = "label = $label",
    params    = { label = label },
  }
end

function KB_Search:search_name(name)
  self.filters[#self.filters + 1] = {
    condition = "name = $name",
    params    = { name = name },
  }
end

function KB_Search:search_property_key(key)
  -- Use jsonb_exists() function instead of ? operator (DBI treats ? as placeholder)
  self.filters[#self.filters + 1] = {
    condition = "jsonb_exists(properties::jsonb, $property_key)",
    params    = { property_key = key },
  }
end

function KB_Search:search_property_value(key, value)
  local json_obj = dkjson.encode({ [key] = value })
  self.filters[#self.filters + 1] = {
    condition = "properties::jsonb @> $json_object::jsonb",
    params    = { json_object = json_obj },
  }
end

function KB_Search:search_starting_path(starting_path)
  self.filters[#self.filters + 1] = {
    condition = "path <@ $starting_path",
    params    = { starting_path = starting_path },
  }
end

function KB_Search:search_path(path_expression)
  self.filters[#self.filters + 1] = {
    condition = "path ~ $path_expr",
    params    = { path_expr = path_expression },
  }
end

function KB_Search:search_has_link()
  self.filters[#self.filters + 1] = {
    condition = "has_link = TRUE",
    params    = {},
  }
end

function KB_Search:search_has_link_mount()
  self.filters[#self.filters + 1] = {
    condition = "has_link_mount = TRUE",
    params    = {},
  }
end

---------------------------------------------------------------------------
-- Query execution
---------------------------------------------------------------------------

local function interpolate(template, params)
  return (template:gsub("%$([%w_]+)", function(name)
    local val = params[name]
    if val == nil then
      error("Missing parameter: " .. name)
    end
    return escape_literal(val)
  end))
end

--- Execute raw SQL, return all rows as array of dicts.
-- Auto-rollbacks on error to prevent "current transaction aborted" cascades.
function KB_Search:_raw_query(sql_str)
  local sth, err = self.dbh:prepare(sql_str)
  if not sth then
    pcall(function() self.dbh:rollback() end)
    error("Prepare error: " .. tostring(err) .. "\nSQL: " .. sql_str)
  end
  local ok, exec_err = sth:execute()
  if not ok then
    pcall(function() self.dbh:rollback() end)
    error("Execute error: " .. tostring(exec_err) .. "\nSQL: " .. sql_str)
  end
  return fetch_all_dicts(sth)
end

--- Execute raw SQL, return single row dict or nil.
-- Auto-rollbacks on error to prevent "current transaction aborted" cascades.
function KB_Search:_raw_query_one(sql_str)
  local sth, err = self.dbh:prepare(sql_str)
  if not sth then
    pcall(function() self.dbh:rollback() end)
    error("Prepare error: " .. tostring(err) .. "\nSQL: " .. sql_str)
  end
  local ok, exec_err = sth:execute()
  if not ok then
    pcall(function() self.dbh:rollback() end)
    error("Execute error: " .. tostring(exec_err) .. "\nSQL: " .. sql_str)
  end
  return fetch_one_dict(sth)
end

--- Commit the current transaction.
function KB_Search:commit()
  self.dbh:commit()
end

--- Rollback the current transaction.
function KB_Search:rollback()
  pcall(function() self.dbh:rollback() end)
end

--- Execute the progressive CTE query with all accumulated filters.
function KB_Search:execute_query()
  if not self.dbh then
    error("Not connected to database. Call new() first.")
  end

  local column_str = "*"

  if #self.filters == 0 then
    local sql_str = string.format("SELECT %s FROM %s", column_str, self.base_table)
    self.results = self:_raw_query(sql_str)
    return self.results
  end

  -- Build CTE chain
  local cte_parts = {}
  cte_parts[1] = string.format("base_data AS (SELECT %s FROM %s)", column_str, self.base_table)

  for i, fi in ipairs(self.filters) do
    local cond   = fi.condition
    local params = fi.params or {}

    local prefixed_params = {}
    local prefixed_cond   = cond
    for pname, pval in pairs(params) do
      local pn = string.format("p%d_%s", i, pname)
      prefixed_params[pn] = pval
      prefixed_cond = prefixed_cond:gsub("%$" .. pname, "$" .. pn)
    end

    local resolved_cond = interpolate(prefixed_cond, prefixed_params)

    local prev_cte = (i == 1) and "base_data" or string.format("filter_%d", i - 1)
    local cte_name = string.format("filter_%d", i)

    if cond and cond ~= "" then
      cte_parts[#cte_parts + 1] = string.format(
        "%s AS (SELECT %s FROM %s WHERE %s)", cte_name, column_str, prev_cte, resolved_cond)
    else
      cte_parts[#cte_parts + 1] = string.format(
        "%s AS (SELECT %s FROM %s)", cte_name, column_str, prev_cte)
    end
  end

  local with_clause  = "WITH " .. table.concat(cte_parts, ",\n")
  local final_select = string.format("SELECT %s FROM filter_%d", column_str, #self.filters)
  local final_query  = with_clause .. "\n" .. final_select

  local ok, res = pcall(self._raw_query, self, final_query)
  if not ok then
    error("Error executing query: " .. tostring(res) .. "\nQuery: " .. final_query)
  end
  self.results = res
  return self.results
end

---------------------------------------------------------------------------
-- Result helpers
---------------------------------------------------------------------------

function KB_Search:get_results()
  return self.results or {}
end

function KB_Search:find_path_values(key_data)
  if not key_data then return {} end
  if key_data.path then key_data = { key_data } end
  local rv = {}
  for _, row in ipairs(key_data) do
    local p = row.path
    if p ~= nil then p = tostring(p) end
    rv[#rv + 1] = p
  end
  return rv
end

function KB_Search:find_description(key_data)
  if not key_data then return {} end
  if key_data.path then key_data = { key_data } end
  local rv = {}
  for _, row in ipairs(key_data) do
    local props = row.properties or {}
    if type(props) == "string" then
      props = dkjson.decode(props) or {}
    end
    local desc = props.description or ""
    local path_key = row.path
    if path_key ~= nil then path_key = tostring(path_key) else path_key = "" end
    rv[#rv + 1] = { [path_key] = desc }
  end
  return rv
end

function KB_Search:find_description_paths(path_array)
  if type(path_array) ~= "table" then
    path_array = { path_array }
  end
  if #path_array == 0 then return {} end

  local placeholders = {}
  for i, p in ipairs(path_array) do
    placeholders[i] = escape_literal(p)
  end

  local sql_str
  if #path_array == 1 then
    sql_str = string.format(
      "SELECT path, data FROM %s WHERE path = %s",
      self.base_table, placeholders[1])
  else
    sql_str = string.format(
      "SELECT path, data FROM %s WHERE path IN (%s)",
      self.base_table, table.concat(placeholders, ","))
  end

  local res = self:_raw_query(sql_str)
  local rv = {}
  local found = {}
  for _, row in ipairs(res) do
    local p = tostring(row.path)
    rv[p] = row.data
    found[p] = true
  end
  for _, p in ipairs(path_array) do
    local ps = tostring(p)
    if not found[ps] then rv[ps] = nil end
  end
  return rv
end

function KB_Search:decode_link_nodes(path)
  if path == nil then
    error("Path must be a non-empty string")
  end
  path = tostring(path)
  if path == "" then
    error("Path must be a non-empty string")
  end

  local parts = {}
  for seg in path:gmatch("[^.]+") do
    parts[#parts + 1] = seg
  end

  if #parts < 3 then
    error(string.format("Path must have at least 3 elements (kb.link.name), got %d", #parts))
  end

  local remaining = #parts - 1
  if remaining % 2 ~= 0 then
    error(string.format(
      "Bad path format: after kb identifier, must have even number of elements, got %d", remaining))
  end

  local kb = parts[1]
  local result = {}
  for i = 2, #parts, 2 do
    result[#result + 1] = { parts[i], parts[i + 1] }
  end
  return kb, result
end

---------------------------------------------------------------------------
-- Expose helpers for sibling modules
---------------------------------------------------------------------------

KB_Search.escape_literal  = escape_literal
KB_Search.fetch_all_dicts = fetch_all_dicts
KB_Search.fetch_one_dict  = fetch_one_dict

return KB_Search